### PR TITLE
fix: single-line flex-container should clamp the line's cross-size

### PR DESCRIPTION
**Pull request by TtTRz**
# Objective

In Taffy, missing a step when calculate the cross size of each flex line. [https://www.w3.org/TR/css-flexbox-1/#algo-cross-line](https://www.w3.org/TR/css-flexbox-1/#algo-cross-line)

``` 
//  If the flex container is single-line, then clamp the line’s cross-size to be within the container’s computed min and max cross sizes.
```

This PR tries to complement this step.

### CODE REVIEW

**Review by nicoburns — CHANGES_REQUESTED** [review]
I've requested a couple of changes, but this generally looks very good: thanks for the PR!

We may also wish to generally review the interaction of:
- The logic added in this PR
- The "if" statement at the top of `calculate_cross_size`
- The presence of min/max sizes (with no preferred size)
- The presence of wrap vs. nowrap
But that does not need to happen in this PR.

File: src/compute/flexbox.rs [hunk 1]
```diff
@@ -1396,6 +1396,16 @@ fn calculate_cross_size(flex_lines: &mut [FlexLine], node_size: Size<Option<f32>
                 })
                 .fold(0.0, |acc, x| acc.max(x));
         }
+        //  If the flex container is single-line, then clamp the line’s cross-size to be within the container’s computed min and max cross sizes.
+        if flex_lines.len() == 1 {
```
[the remark this answers is not in the document]
Comment (TtTRz) [reply src/compute/flexbox.rs:1400]: I noticed that `flex_lines.len() == 1` is used in [https://github.com/DioxusLabs/taffy/blob/main/src/compute/flexbox.rs#L1352](https://github.com/DioxusLabs/taffy/blob/main/src/compute/flexbox.rs#L1352) to determine if it is a single line. 
I've also looked at some other browser layout engine implementations, I see no difference between "nowrap" and "happends to be single-line", but i'll add some tests to check it.

[the remark this answers is not in the document]
Comment (nicoburns) [reply src/compute/flexbox.rs:1400]: > I've also looked at some other browser layout engine implementations, I see no difference between "nowrap" and "happends to be single-line", but i'll add some tests to check it.

If you add `flex-wrap: wrap` (the default is `nowrap`) to the test you created for this PR then you'll find that the layout ends up changing. I rather suspect that line 1532 is wrong too (although that `if` does also include the flex_wrap property later in the condition).

[the remark this answers is not in the document]
Comment (TtTRz) [reply src/compute/flexbox.rs]: I found the difference just like you said, i think we should use flex_wrap to determine if it is single-line or not.

I add some test cases and it looks fine :)
* padding/border
* min/max size
* wrap/no-wrap

File: src/compute/flexbox.rs [hunk 2]
```diff
@@ -1396,6 +1396,16 @@ fn calculate_cross_size(flex_lines: &mut [FlexLine], node_size: Size<Option<f32>
...
+            let cross_axis_padding_border = constants.content_box_inset.cross_axis_sum(constants.dir);
+            let cross_min_size = constants.min_size.cross(constants.dir);
+            let cross_max_size = constants.max_size.cross(constants.dir);
+            flex_lines[0].cross_size = flex_lines[0].cross_size.maybe_clamp(
+                cross_min_size.maybe_sub(cross_axis_padding_border),
+                cross_max_size.maybe_sub(cross_axis_padding_border),
```
[the remark this answers is not in the document]
Comment (TtTRz) [reply src/compute/flexbox.rs:1406]: I added test cases

**Review by nicoburns — APPROVED** [review]
This looks good to me now. Thanks for the fixes (and the tests)!

**Comment by TtTRz** [comment]
> This looks good to me now. Thanks for the fixes (and the tests)!

Thanks for the review XD!

**Pull request merged**